src/support/hazard.c

```diff
@@ -102,6 +98,13 @@ __wt_hazard_set_func(WT_SESSION_IMPL *session, WT_REF *ref, bool *busyp
         WT_ASSERT(session,
           session->nhazard == session->hazard_inuse &&
             session->hazard_inuse < session->hazard_size);
+        /*
+         * Our session->hazard pointer may have been updated by hazard_grow. If this is the case we
+         * need to ensure the updated pointer is visible to other threads before the increment of
+         * hazard_inuse as we use this value to walk the hazard array, and using the new
+         * hazard_inuse on the old hazard array can lead to reading past the end of the array.
+         */
+        WT_WRITE_BARRIER();
```

Comment:
A `WRITE_BARRIER` seemed cleaner than a `PUBLISH` here as the `PUBLISH` ends up looking like
```
WT_PUBLISH(session->hazard_inuse, session->hazard_inuse + 1);
 hp = &session->hazard[session->hazard_inuse];
```
Open to discussion on this one.

Reply:
Wouldn't this be:
```
hp = &session->hazard[session->hazard_inuse];
WT_PUBLISH(session->hazard_inuse, session->hazard_inuse + 1);
```
It's a post increment originally.

Reply:
Yes is would. My focus was more on the 3 separate uses of `session->hazard` if we use `WT_PUBLISH` over a `WT_WRITE_BARRIER`

Reply:
So is this the `Ensure Invariant` usage pattern? I am struggling to see the exact ordering.

Reply:
I think I get it after reading the usage patterns doc, we need the increment to happen after the grow to guarantee a reader doesn't get an array index OOB error.

Reply:
Yep. On the writer thread's side `session_inuse` will always be in bounds for the new array, so we just need to ensure that the new array is visible before an increment of `session_inuse`. This only applies in the case we need to increment the array.

Reply:
Wouldn't it be better placed in the grow function then? As that function has the context as to why we need that ordering.

Reply:
I'm open to it. 
By doing that we lose the context of exactly which store is being ordered; When I first looked at this code I started with the ordered reads in `hazard_get_reference` and was confused by the lack of write barriers associated with the `hazard_inuse++`. I think our struggle is the logic needing ordering exists across both `hazard_set_func` and its subfunction `hazard_grow`.

I'll stare at this for a bit.

Reply:
I would put it in grow with a comment indicating which var it is ordering against, and a comment in hazard set explaining that there is ordering that it depends on.

Reply:
Updated in https://github.com/wiredtiger/wiredtiger/pull/9530/commits/c839f94a6c8183633b1749ef93ab8b95900aeca5